File: week3/p3.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* next;
};

struct Node* newNode(int data) {
    struct Node* node = (struct Node*)malloc(sizeof(struct Node));
    node->data = data;
    node->next = NULL;
    return node;
}

struct Node* deleteNode(struct Node* head, int x) {
    if (head == NULL) return NULL;

    if (x == 1) {
        struct Node* temp = head;
        head = head->next;
        free(temp);
        return head;
    }

    struct Node* prev = head;
    for (int i = 1; i < x - 1 && prev != NULL; i++)
        prev = prev->next;

    if (prev != NULL && prev->next != NULL) {
        struct Node* temp = prev->next;
        prev->next = temp->next;
        free(temp);
    }
    return head;
}

void printList(struct Node* head) {
    while (head != NULL) {
        printf("%d ", head->data);
        head = head->next;
    }
    printf("\n");
}

int main() {
    struct Node* head = newNode(1);
    head->next = newNode(3);
    head->next->next = newNode(4);

    head = deleteNode(head, 3);
    printList(head);
    return 0;
}

File: week3/p3.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

def deleteNode(head, x):
    if not head:
        return None

    if x == 1:
        return head.next

    prev = head
    for _ in range(x - 2):
        if prev is None:
            return head
        prev = prev.next

    if prev and prev.next:
        prev.next = prev.next.next

    return head

def printList(head):
    temp = head
    while temp:
        print(temp.data, end=" ")
        temp = temp.next
    print()

# Example
head = Node(1)
head.next = Node(3)
head.next.next = Node(4)

head = deleteNode(head, 3)
printList(head)
